### PR TITLE
add positional records

### LINKED ISSUE #21: Add support for records
First of all... great tool! Many thanks.

It would be great if the CodeGenHelper is able to support records.
Are there any plans?

### PR DESCRIPTION
**Pull request by MichaelHochriegl**
### Description
Implemented a positional based record builder.

### Example
```csharp//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

namespace AwesomeApp
{
    public SampleRecord(string FirstName, string LastName);
}
```

Closes #21

### CODE REVIEW

File: tests/CodeGenHelpers.Tests/SampleCode/MyComplexClass.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace CodeGenHelpers.SampleCode
+{
+    public class MyComplexClass
+    {
+
+    }
```
Comment (dansiegel) [inline tests/CodeGenHelpers.Tests/SampleCode/MyComplexClass.cs:6]: This doesn't appear to be used or complex 😉 

The SampleCode is used in-lieu of strings in the tests in order to keep the tests themselves cleaner.

**Review by dansiegel — CHANGES_REQUESTED** [review]
While this does support the first use case with Records I really want to be sure that we are supporting both cases for Records with this API. 

Case 1:
You have Parameters and no Properties
```cs
public record Foo(string Bar)
```

Case 2:
You have Properties and no parameters
```cs
public record Foo
{
    public string Bar { get; init; }
}
```

I believe what we're really looking for is going to be something more like the following. We can default to the first case and expose a helper method to control the output style. We want to think of these as Properties on the Record rather than Parameters of the constructor. How we initialize those properties is ultimately determined by whether or not we've called `UseInitProperties()`
```cs
CodeBuilder.Create("AwesomeApp")
    .AddRecord("SampleRecord")
      .UseInitProperties()
      .AddProperty("string", "FirstName");
```

**Comment by dansiegel** [comment]
you may want to rebase this and update to use the new TestBase class that will help you to more easily reference sample files for the expected output.

File: tests/CodeGenHelpers.Tests/SampleCode/MyComplexClass.cs (hunk 1 above)
Comment (MichaelHochriegl) [reply tests/CodeGenHelpers.Tests/SampleCode/MyComplexClass.cs:6]: Sorry, this slipped in and clearly shouldn't be there :sweat_smile:. I only used it to do some local tests to make sure I didn't change any existing behaviour.

**Comment by MichaelHochriegl** [comment]
Thanks for the feedback! Will work on this this evening and apply the changes accordingly :thumbsup:

File: src/CodeGenHelpers/RecordPositionalBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            _xmlDoc.Write(writer);
+
+            foreach (var attribute in _attributes)
+                writer.AppendLine($"[{attribute}]");
+
+            var modifier = AccessModifier switch
+            {
+                null => "public",
+                _ => AccessModifier.ToString().ToLower()
+            };
```
[the remark this answers is not in the document]
Comment (MichaelHochriegl) [reply src/CodeGenHelpers/RecordPositionalBuilder.cs:102]: Sorry for bothering you again, I guess if you have implemented it yourself you would have been already done by now :sweat_smile:

I just put some thought into this dual-support for both types of records and would like to get your input on this.
Should this be done using a new PropertyBuilder (e.g. RecordPropertyBuilder) that will produce both outputs according to the mode or would you prefer an approach that still uses the current PropertyBuilder for the property record and the ParameterBuilder for the parameter record?

Or am I completely on the wrong side with my current thoughts?
(Sorry, first time contributing to open source)

**Comment by dansiegel** [comment]
@MichaelHochriegl just wanted to follow up on this. Would love to get this merged so I can get a new release out

**Comment by MichaelHochriegl** [comment]
Sorry for taking so long, Dan. Did you get a chance to think about my questions posted above?

> Sorry for bothering you again, I guess if you have implemented it yourself you would have been already done by now sweat_smile
> 
> I just put some thought into this dual-support for both types of records and would like to get your input on this.
> Should this be done using a new PropertyBuilder (e.g. RecordPropertyBuilder) that will produce both outputs according to the mode or would you prefer an approach that still uses the current PropertyBuilder for the property record and the ParameterBuilder for the parameter record?
> 
> Or am I completely on the wrong side with my current thoughts?
> (Sorry, first time contributing to open source)

I'm still kinda torn on which implementation would fit the library best. If you have any input on this I will start to work on it ASAP.

**Comment by dansiegel** [comment]
Hopefully this helps.... I believe the API we want is something along the lines of:

```cs
var builder = CodeBuilder.Create("AwesomeApp")
    .AddRecord("MyRecord")
        .AddProperty(...)
```

We might have a builder like:

```cs
public enum RecordPropertyType
{
    Positional,
    Init
}

public class RecordBuilder
{
    internal RecordPropertyType DefaultPropertyType { get; private set; }
    
    public RecordBuilder UsePositionalProperties() => this;
    
    public RecordBuilder UseInitProperties() => this;
    
    public RecordPropertyBuilder AddProperty(string name, RecordPropertyType? type)
    {
        var prop = new RecordPropertyBuilder(name, type, this);
        _properties.Add(prop);
        return prop;
    }
}
```

**Comment by MichaelHochriegl** [comment]
Sorry for taking so long, didn't get the notification that you replied. I have the positional version restructured and tested with the TestBase approach.
I'm currently at implementing the init-approach and noticed that there is a race-condition problem with the writer.

I will open a new issue for this, as it is not really related to the RecordBuilder, as it also happens to the ClassBuilder.

**Comment by dansiegel** [comment]
1.5.6 is now available on nuget.org with the support for Records. Thanks again for all of the hard work on this PR.

**Pull request closed without merging**